doc/dev/crush-msr.rst

```diff
@@ -0,0 +1,140 @@
+============================
+CRUSH MSR (Multi-step Retry)
+============================
+
+Motivation
+----------
+
+Conventional CRUSH has an important limitation: rules with
+multiple choose steps which hit an out osd cannot retry
+prior steps.  As an example, with a rule like
+::
+
+    rule replicated_rule_1 {
+        ...
+        step take default class hdd
+        step chooseleaf firstn 3 type host
+        step emit
+    }
+
+one might expect that if all of the osds on a particular host
+are marked out, mappings including those OSDs would end up
+on another host (provided that there are enough hosts).  Indeed,
+that's what will happen.  Moreover, if 1/8 OSDs on a host are
+marked out, roughly 1/8 of the PGs mapped to that host will end
+up remapped to some other host keeping overall per-OSD utilization
+even.
+
+Suppose, instead, the rule were written like this:
+::
+
+    rule replicated_rule_1 {
+        ...
+        step take default class hdd
+        step choose firstn 3 type host
+        step choose firstn 1 type osd
+        step emit
+    }
+
+The behavior would be very similar as long as no OSDs are marked
+out.  However, if an OSD is marked out, any PGs mapped to that
+OSD will be remapped to other OSDs on the same host resulting in
+those OSDs being over-utilized relative to OSDs on other hosts.
+Moreover, if all of the OSDs on a host are marked out, mappings
+that happen to hit that host will fail resulting in undersized PGs.
+
+As long as the goal is to split N OSDs between N failure domains,
+the solution is simply to use the chooseleaf variant above.  However,
+consider a use case where we want to split an 8+6 EC code over 4
```

Comment:
s/code/stripe/

Reply:
Not stripe.  I'll change it to encoding.